indexing-service/src/main/java/io/druid/indexing/common/task/CompactionTask.java

```diff
@@ -263,34 +282,75 @@ static IndexIngestionSpec createIngestionSchema(
     }
 
-    final DataSchema dataSchema = createDataSchema(
-        segmentProvider.dataSource,
-        segmentProvider.interval,
-        dimensionsSpec,
-        toolbox.getIndexIO(),
-        jsonMapper,
-        timelineSegments,
-        segmentFileMap
-    );
-    return new IndexIngestionSpec(
-        dataSchema,
-        new IndexIOConfig(
-            new IngestSegmentFirehoseFactory(
-                segmentProvider.dataSource,
-                segmentProvider.interval,
-                null, // no filter
-                // set dimensions and metrics names to make sure that the generated dataSchema is used for the firehose
-                dataSchema.getParser().getParseSpec().getDimensionsSpec().getDimensionNames(),
-                Arrays.stream(dataSchema.getAggregators()).map(AggregatorFactory::getName).collect(Collectors.toList()),
-                toolbox.getIndexIO()
-            ),
-            false
-        ),
-        tuningConfig
-    );
+    if (keepSegmentGranularity) {
+      // if keepSegmentGranularity = true, create indexIngestionSpec per segment interval, so that we can run an index
+      // task per segment interval.
+      final List<IndexIngestionSpec> specs = new ArrayList<>(timelineSegments.size());
+      for (TimelineObjectHolder<String, DataSegment> holder : timelineSegments) {
+        final DataSchema dataSchema = createDataSchema(
+            segmentProvider.dataSource,
+            holder.getInterval(),
+            Collections.singletonList(holder),
+            dimensionsSpec,
+            toolbox.getIndexIO(),
+            jsonMapper,
+            segmentFileMap
+        );
+
+        specs.add(
+            new IndexIngestionSpec(
+                dataSchema,
+                new IndexIOConfig(
+                    new IngestSegmentFirehoseFactory(
+                        segmentProvider.dataSource,
+                        holder.getInterval(),
+                        null, // no filter
+                        // set dimensions and metrics names to make sure that the generated dataSchema is used for the firehose
+                        dataSchema.getParser().getParseSpec().getDimensionsSpec().getDimensionNames(),
+                        Arrays.stream(dataSchema.getAggregators()).map(AggregatorFactory::getName).collect(Collectors.toList()),
+                        toolbox.getIndexIO()
+                    ),
+                    false
+                ),
+                tuningConfig
+            )
+        );
+      }
+
+      return specs;
+    } else {
+      final DataSchema dataSchema = createDataSchema(
+          segmentProvider.dataSource,
+          segmentProvider.interval,
+          timelineSegments,
+          dimensionsSpec,
+          toolbox.getIndexIO(),
+          jsonMapper,
+          segmentFileMap
+      );
+
+      return Collections.singletonList(
+          new IndexIngestionSpec(
+              dataSchema,
+              new IndexIOConfig(
```

Comment:
maybe worth pulling out a method or lambda fn maybe `IndexIOConfig makeConfig(dataSchema, interval)` since constructor statements are pretty big and largely the same?

Reply:
Sounds good. Fixed.